♻️ Separate LCD pins from board pins

thinkyhead opened the pull request:
Gradual migration of LCDs to separate pins file(s) using defined connectors / interfaces.

GMagician commented:
I'm not using anymore AGCM4 + RAMPS_144 but it seems that on AUX1 you have:
  VCC GND D1 D0
  VCC GND A3 A4
but VCC may be 5V/3V depending on J4 jumper (this goes also to IO pin)

AUX3 is (please note that SPI pins on AGCM4 are not like DUE but standard D50-52 may be redirected via solering pads):
 RESET GND SCK MISO VCC
     NC    5V D53 MOSI D49

AUX4 is:
  VCC and not 5V

GMagician commented:
[RAMPS_1.4.4_C6.pdf](https://github.com/MarlinFirmware/Marlin/files/11183286/RAMPS_1.4.4_C6.pdf)
[LCD144-A_Rev.A.pdf](https://github.com/MarlinFirmware/Marlin/files/11183288/LCD144-A_Rev.A.pdf)

As reference I add schematic

thinkyhead committed:
♻️ Migrate IS_RRD_FG_SC

thinkyhead committed:
♻️ Migrate IS_RRD_FG_SC

thinkyhead committed:
♻️ Migrate ENDER2_STOCKDISPLAY

thinkyhead committed:
♻️ Migrate ENDER2_STOCKDISPLAY

thinkyhead committed:
♻️ Migrate RRD SC

thinkyhead committed:
♻️ Migrate RRD SC

thinkyhead committed:
♻️ Migrate ULTI_CONTROLLER

thinkyhead committed:
♻️ Migrate ULTI_CONTROLLER

thinkyhead committed:
♻️ Migrate CR10_STOCKDISPLAY

thinkyhead committed:
♻️ Migrate CR10_STOCKDISPLAY

thinkyhead committed:
♻️ Migrate MINIPANEL

thinkyhead committed:
♻️ Migrate MINIPANEL

GMagician commented on Marlin/src/pins/samd/pins_RAMPS_144.h:
```diff
@@ -260,79 +275,224 @@
+#define AUX3_06                               53
+
+//
+// AUX4    5V GND D32 D47 D45 D43 D41 D39 D37 D35 D33 D31 D29 D27 D25 D23 D17 D16
```
VCC GND ....

GMagician commented on Marlin/src/pins/samd/pins_RAMPS_144.h:
```diff
@@ -260,79 +275,224 @@
+//          2   4   6   8
+//          1   3   5   7
+//         5V  GND A3  A4
+//
```
AUX1 is:
  VCC GND D2 D1
  VCC GND A3 A4

GMagician commented on Marlin/src/pins/samd/pins_RAMPS_144.h:
```diff
@@ -260,79 +275,224 @@
+//          7   5   3   1
+//          8   6   4   2
+//         NC  D53 D51 D49
+//
```
AUX3 in ramps 1.4.4 is a 5x2 connector
  RST GND  SCK MISO VCC
  NC  5V     D53 MOSI D49

because D52, D50 and D51 are not SPI pins, real SPI pins are named SCK, MISO and MOSI and are the same pins used by SPI connector on due (named ICSP on AGCM4)

GMagician replied:
Even if ramps 1.4.4 supports due and then may redirect D52, D50 and D51 on AUX3 this pinout is specific for AGCM then it requires such changes

thinkyhead replied:
"RAMPS 1.4.4" by itself is meaningless. It requires a base mapping of the underlying pins that mate up with the shield. The basic `pins_RAMPS.h` is based on a RAMPS 1.3/1.4 mated with a standard Arduino Mega.

GMagician replied:
on this assumptions you have to move it outside samd folder, change board name (AGCM4 RAMPS 1.4.4) and remove:
```
#if NOT_TARGET(ARDUINO_GRAND_CENTRAL_M4)
  #error "Oops! Select 'Adafruit Grand Central M4' in 'Tools > Board.'"
#endif
```

GMagician replied:
but doing this it will be no more useable with an AGCM4. A new pins definition should be done, derived from this, with correct SPI changes

thinkyhead replied:
> you have to move it outside samd folder

If RAMPS 144 is the only way that AGCM4 can be interfaced with Marlin, then the file might as well be renamed to `pins_AGCM4_RAMPS_144.h`. If AGCM4 always uses the SAMD processor then it should remain in the "samd" folder. If other boards arrive that want to interface with RAMPS 144, they can be named to indicate them as such.

The longstanding use of a "ramps" folder to contain boards that follow the RAMPS pin mappings might just be approaching obsolescence too. It has been convenient mainly due to the shared remapping of Extruder / Bed / Fan pins based on configuration. That might also get an additional layer similar to this LCD refactor.

Each of these tasks is a small pile of work, so be patient. I will continue on all of these concepts alonside my regular duties.

GMagician replied:
Well at present this is the only shield that can be used by AGCM4 (Adafruit Grand Central M4), this board mounts only SAMD51.
As said before ramps 144 has been discontinued so it won't be used anymore (it never left beta testing afaik).
Not sure how many people are using it, I suspect I'm the only one since no one posted a pins.h before (then no "due" user I think). I also really think I'm the only one that used AGCM4 on Marlin (now dropped in favor of S6).

|Each of these tasks is a small pile of work, so be patient
No reason to say that. I know the big amount of tasks you are involved with Marlin so take your time, thanks

thinkyhead commented:
If this isn't breaking anything I may go ahead and merge this incomplete LCD refactor to get the ball rolling. This set of changes will be very important to the continued relevance and viability of Marlin Firmware going forward.

ellensp commented:
please no!

128x64 lcds and 20x4 displays are already broken, with moves refactors and spacing issues
touch screen are all now broken after latest touch refactors
SPI TFT displays have been broken for a long time on many platforms.

Please don't break any more displays any further than they are already.

thinkyhead commented:
> 128x64 lcds and 20x4 displays are already broken, with spacing issues
> marlin controlled touch screen are all now broken after latest touch refactors
> SPI TFT displays have been broken for a long time on many platforms

That is all related to code logic, not to pins. If this PR is not changing the pins that are assigned for LCDs (or otherwise making those issues worse) then these other issues have no bearing on this PR. We can speculate about whether this PR would affect those issues, but without any data that is just speculation.

Thanks for bringing those issues up, as it is a major priority with this next release to fix up all remaining and known bugs. So let's keep plugging away at fixing these things and see if we can locate more nerds who like to dive in and solve these issues so they don't linger for such a long time.

tombrazier commented:
Scott I guess the rationale for this change is to remove duplication with the same displays having to be defined in the pins file of every mainboard they are compatible with. Is that correct?

Looking at the code diffs, the logic of the change seems simple enough but the scale is immense. With that much copying I think the likelihood of an error is high. On the other hand, I prefer smaller merges more often to larger merges more seldom so if the only choice is between merging now or doing even more changes and merging later then now is better. Even better would be if this could be broken down into smaller components.

I am mindful of @ellensp's concerns as well. Better to get displays working generally before refactoring.

How much testing has this PR had? Does it even compile for every potential combination?

sjasonsmith commented:
When complete this is going to be a good change. Maintaining every screen in every pins file is never-ending. It's also difficult to understand if they are correct when pins are added.

I don't think this is ready to go in. I see lots of details moved to pins/lcd files, but I don't see anything actually _including_ those files yet. I also don't think there would be any practical way for me to review the changes and actually confirm they are correct.


The following is what I think could be a safe way to approach this with minimal disruption for anyone:

1. Stage and commit all of the non-functional changes. This touches a lot of comments which add noise and make it harder to spot which changes in the PR deserve attention. The smaller the scope of each PR, the easier it is for someone to volunteer to review and test those changes, _and_ actually understand what they're looking at.

2. Write a script which will set a board (one per pins file) and use the preprocessor to dump all the symbols it generates. This should be repeated for each screen which is being extracted into a new file. We have done something similar to this in the past when we were playing with adding embedded configurations, so I don't think this should be too hard.

3. Add the new LCD pins files, but ONLY include them if the board pins file has "opted-in" by declaring something like USE_COMMON_LCD_PINS. This flag could be removed once the transition is complete.

4. As pins files opt-in, re-run the script from step 2 and compare the output before/after output to check for unexpected differences. Fixing them at this stage will be a lot easier than figuring out what's wrong after people are using the change.


I'm sure there will still be some mistakes made...especially in the "weird' boards with non-standard connectors, but this should enable the majority to be transitioned without requiring any specific hardware testing.

sjasonsmith commented:
I'm not really going to be available today, but I'm willing to contribute to creating that script I mentioned in step 2.

tombrazier commented:
I support everything @sjasonsmith says.

sjasonsmith commented on Marlin/src/pins/sanguino/pins_MELZI_CREALITY.h:
```diff
@@ -38,122 +38,102 @@
 
 #define BOARD_INFO_NAME "Melzi (Creality)"
```
MELZI_CREALITY is seeing some pin changes when using CR10_STOCKDISPLAY, which is probably the most common display used with this board.

Old:
```
#define BEEPER_PIN EXP1_01_PIN // (Evaluates to 27)
```

New:
```
#define BEEPER_PIN -1 // Should probably be AUX1_01. This probably was removed because it is frequently repurposed for SERVO0 for BLTouch.
```

sjasonsmith replied:
It looks like the BEEPER change may be related to an incompletely implementation in adapters.h:

```
#elif ENABLED(ADAPTER_CREALITY_PIN27)

  // Creality or BTT "Pin 27" Adapter with BLTouch pins
  ```

thinkyhead replied:
I fixed `BEEPER_PIN` with this change to `pins_MELZI_CREALITY.h`…
```diff
- #if SERVO0_PIN == BEEPER_PIN || FIL_RUNOUT_PIN == BEEPER_PIN
+ #if PIN_EXISTS(BEEPER) && (SERVO0_PIN == BEEPER_PIN || FIL_RUNOUT_PIN == BEEPER_PIN)
```

thinkyhead replied:
> … may be related to an incompletely implementation in `adapters.h`:

Not in this case, at least not yet. A "Pin 27" adapter applied to any board's EXP1 connector(s) will make the LCD ignore pin 27 so that a probe or filament sensor can use it. The simplest handling for this would be to add to `adapters.h` above the "set undefined to `-1`" section…
```cpp
#if ENABLED(ADAPTER_PIN27)
  #undef LCD1_01_PIN
#endif
```

We could also look for situations where EXP1_01 is being used for a probe or filament sensor in `adapters.h` or `pins_lcd.h` and alert the user that they need some kind of "Pin 27" adapter. (So users will sometimes need to tell us about their adapter in addition to their LCD after this refactor.)

sjasonsmith commented on Marlin/src/pins/lpc1768/pins_BTT_SKR_V1_3.h:
```diff
@@ -220,47 +220,21 @@
 #define EXP2_07_PIN                        P1_31
 #define EXP2_08_PIN                        -1
 
-#if HAS_WIRED_LCD
```
SKR 1.3 when using CR10_STOCKDISPLAY now sets a KILL_PIN, which it previously did not. I don't recall wither this display supports a reset button or not.

Old:
`#define KILL_PIN -1`

New:
`#define KILL_PIN LCD1_04_PIN // Maps to P1_30`

sjasonsmith replied:
I saw this same behavior with CR10_STOCKDISPLAY on other boards as well.

ellensp replied:
no reset button visible https://marlinfw.org/docs/hardware/controllers.html#cr10_stockdisplay
But things like btt tft35 have a reset button and emulate a cr10_stockdisplay
So its probably good to define a kill pin.

thisiskeithb replied:
> But things like btt tft35 have a reset button and emulate a cr10_stockdisplay
> So its probably good to define a kill pin.

Ahh, yes. I forgot about that. They have a reset button & it works!

<img width="688" alt="image" src="https://github.com/MarlinFirmware/Marlin/assets/13375512/cd076738-c362-405c-b508-9b0148419efc">

ellensp replied:
Also the [CR10_schematic](https://github.com/RudolphRiedel/CR-10_wiring/blob/master/Ender3_schematic.PDFl)

shows the following, with a reset pin defined. 

<img alt="Screen Shot 2019-08-19 at 9 41 40 PM" width="353" src="https://user-images.githubusercontent.com/13375512/63318128-34692880-c2ca-11e9-9fb3-e8087ae3290f.png">

thinkyhead replied:
With the new adapters and individual display definitions we'll get the correct result more often. For boards that have a RESET line on the EXP port the display reset button will get mapped to it, and if they don't then it will just be mapped to -1.

sjasonsmith commented on Marlin/src/pins/stm32f1/pins_BTT_SKR_MINI_E3_common.h:
```diff
@@ -124,18 +124,35 @@
-
-#if HAS_DWIN_E3V2 || IS_DWIN_MARLINUI
+
+/*   -----
+ *   | 1 | RST
+ *   | 2 | PA3 RX2
+ *   | 3 | PA2 TX2
+ *   | 4 | GND
+ *   | 5 | 5V
+ *   -----
+ *    TFT
+ */
+
+#define TFT_02                              PA3
+#define TFT_03                              PA2
```
(Also applies to E3 DIP board)

I don't really like these names `TFT_02` and `TFT_03`. I see that these are just documenting pin numbers in the connector on the board, but the `TFT_` prefix is used by many TFT-related configuration options as well.

Perhaps this should be more clearly scoped to the board, to avoid confusing readers into thinking these are consumed directly by TFT code elsewhere.

Maybe `BOARD_TFT_02`?

thinkyhead replied:
On another note, this is a serial port so I assume it's used as such, so we may just end up using the define `LCD_SERIAL_PORT` for displays like this and define a dummy "adapter" that simply indicates the selected LCD needs a serial port.

sjasonsmith commented on Marlin/src/pins/mega/pins_MEGACONTROLLER.h:
```diff
@@ -129,35 +129,50 @@
-
+  // Migrated to pins/lcd
+  // Real MINIPANEL is always mapped to LCD1/2
+  #define ADAPTER_MINIPANEL_DUAL
```
MEGACONTROLLER seems to have quite a few changes when used with MKS_MINI_12864, which enables the ADAPTER_MINIPANEL_DUAL adapter. I'm not sure which of these were actually used in this case, versus losing functionality.

- KILL_PIN is now -1, was previously 12.
- LCD_BACKLIGHT_PIN is now missing, was previously 44
- SD_DETECT_PIN is now -1, was previously 49

-

thinkyhead replied:
Lots of good information in this and other comments, thanks! I'll rebase this PR shortly, then I'll aim to check all the notes and apply any needed adjustments. I don't want to keep piling more displays into this PR until what we have so far is working and not causing any pins to be defined wrongly. I'd like to merge this once it is bug-free, but I know that's a tricky thing to verify. Some of these changes actually fixed a thing or two, so that's a good sign. I do believe this is a good direction to go. I'll take all notes above under advisement!

sjasonsmith replied:
There may be other impacted boards as well, I haven’t gone through everything. I figured I had posted enough for the next round, and can revisit it again when ready.

thinkyhead replied:
> `MEGACONTROLLER` seems to have quite a few changes when used with `MKS_MINI_12864`, which enables the `ADAPTER_MINIPANEL_DUAL`

Oops. Forgot to include in `pins_lcd.h`…
```cpp
  #elif ENABLED(MINIPANEL)
    #include "lcd/MINIPANEL.h"
```
Everything else looks correct, but I will triple-check.

sjasonsmith reviewed (commented):
I've spot-checked the #define output of quite a few common board/display combinations, and things are actually looking pretty good.

I've found a few anomalies so far, but I ran out of time to review more right now. I'm currently working my way down adapters.h testing those, as they seem to be more problematic than the direct display connections.

thinkyhead commented:
> I don't see anything actually _including_ those files yet.

You probably already figured this out, but the appropriate LCD defines are included automagically by `pins_lcd.h` which is included at the end of `pins.h`…
```cpp
//
// LCD / Controller Pins based on EXP ports
//
#include "pins_lcd.h"
```

thinkyhead committed:
♻️ Migrate FYSETC_MINI_12864

thinkyhead committed:
♻️ Migrate FYSETC_MINI_12864

thinkyhead committed:
♻️ Migrate REPRAPWORLD_GRAPHICAL_LCD

thinkyhead committed:
♻️ Migrate REPRAPWORLD_GRAPHICAL_LCD

sl1pkn07 commented:
Hello

in #26719 you swapped the BNT buttons in the ReARM board for the CR10_STOCK_DISPLAY

can ask why?

greetings

thinkyhead commented:
> in #26719 you swapped the BNT buttons in the ReARM board for the `CR10_STOCK_DISPLAY`
> can ask why?

To eliminate this pointless requirement only mentioned in the pins file:
```cpp
// Requires REVERSE_ENCODER_DIRECTION in Configuration.h
```

sl1pkn07 commented:
oh. i forgot that

thanks

thinkyhead committed:
♻️ (Migrate YHCB2004)

thinkyhead committed:
♻️ (Migrate YHCB2004)

thinkyhead committed:
♻️ LCD Adapters

thinkyhead committed:
♻️ LCD Adapters

thinkyhead committed:
♻️ Migrate WYH_L12864

thinkyhead committed:
♻️ Migrate WYH_L12864

thinkyhead committed:
♻️ Migrate HAS_WIRED_LCD (fallback)

thinkyhead committed:
♻️ Migrate HAS_WIRED_LCD (fallback)

thinkyhead committed:
♻️ Migrate MKS_MINI_12864

thinkyhead committed:
♻️ Migrate MKS_MINI_12864

thinkyhead committed:
♻️ Migrate FYSETC_242_OLED_12864

thinkyhead committed:
♻️ Migrate FYSETC_242_OLED_12864

thinkyhead committed:
♻️ Migrate ANET_FULL_GRAPHICS_LCD

thinkyhead committed:
♻️ Migrate ANET_FULL_GRAPHICS_LCD

thinkyhead committed:
Remove migrated CTC_A10S_A13 inline LCD block from pins_ANET_10

CTC_A10S_A13 pins now come from lcd/ANET_FULL_GRAPHICS_LCD.h.

ANET_FULL_GRAPHICS_LCD is kept inline because the ANET 1.0 board uses
a non-standard EXP1 layout: LCD_PINS_D4 is on EXP1_02 (not EXP1_04),
freeing EXP1_04 for SERVO0 (BLTouch/3D-Touch). Board-specific ST7920
timing delays also differ from the generic controller file.

thinkyhead committed:
Document CR10_STOCKDISPLAY inline retention in pins_MELZI_MALYAN

The Malyan Melzi board uses direct ATmega pin numbers for its LCD wiring
rather than EXP connector tokens, so LCD1_xx aliases are unavailable and
the block cannot be migrated to lcd/CR10_STOCKDISPLAY.h.

All other boards with CR10_STOCKDISPLAY blocks have been verified:
- Already migrated (stub-only): 38 files across esp32, hc32f4, lpc1768/9,
  mega, native, ramps, rp2040, sam, samd, sanguino, stm32f1/f4/g0/h7
- Kept inline with board-specific BEEPER_PIN=-1 guards (USE_PIN_27_BOARD):
  hc32f4/pins_AQUILA_101, stm32f1/pins_CREALITY_V4,
  stm32f4/pins_CREALITY_CR4NTXXC10 (handled by #ifndef in lcd/ file)

thinkyhead committed:
Remove redundant KILL_PIN/NEOPIXEL from FLY_D8 FYSETC_242_OLED_12864 block

KILL_PIN -1 and BOARD_NEOPIXEL_PIN are already defined in
lcd/FYSETC_242_OLED_12864.h (KILL_PIN=-1, NEOPIXEL=LCD1_07_PIN=EXP1_07_PIN).
Retain FORCE_SOFT_SPI as a board-specific requirement.

All other boards with FYSETC_242_OLED_12864 blocks were already clean.

thinkyhead committed:
Document MINIPANEL inline retention: non-EXP wiring in 9 board files

All 9 files use board-specific raw pin numbers or AUX2/STM32/Teensy
connectors with no EXP1/EXP2 mapping — lcd/MINIPANEL.h tokens cannot
apply. Added NOTE comments to each MINIPANEL block explaining why.

Boards that use only adapter flags (MELZI_CREALITY) or #error guards
(STM32F1R) are already correct and need no change.

No pin definitions removed.

thinkyhead committed:
Add ADAPTER_RAMPS_AUX2 / DUO_AUX2 adapters; wire MINIPANEL boards to EXP tokens

- lcd/adapters.h: Add ADAPTER_RAMPS_AUX2 (RAMPS AUX-2 connector pinout)
  and ADAPTER_RAMPS_DUO_AUX2 (same with 4 Due-specific pin overrides)
- pins_RAMPS.h, pins_RAMPS_NATIVE.h, pins_TT_OSCAR.h, pins_RP2040.h:
  Replace inline MINIPANEL defines with ADAPTER_RAMPS_AUX2 token;
  lcd/adapters.h now supplies LCD1/LCD2 mappings for lcd/MINIPANEL.h
- pins_RAMPS_DUO.h: Replace undef/override pattern with ADAPTER_RAMPS_DUO_AUX2
- pins_WANHAO_ONEPLUS.h, pins_DUE3DOM_MINI.h, pins_CHITU3D.h,
  pins_PRINTRBOARD_REVF.h: Replace inline MINIPANEL defines with EXP1/EXP2
  pin definitions; default adapters.h passthrough routes LCD1/LCD2 -> EXP1/EXP2,
  enabling lcd/MINIPANEL.h to supply the final signal assignments

thinkyhead committed:
pins: LCD RRD_FG_SC sweep — notes + Smoothieboard Technique C

- pins_SELENA_COMPACT.h: add NOTE (non-standard D5-D7 layout, no EXP tokens)
- pins_SMOOTHIEBOARD.h: Technique C — add EXP1/EXP2 defs matching standard LPC1769
  layout; remove inline IS_RRD_FG_SC pin defines, now handled by lcd/RRD_FG_SC.h
- pins_MIGHTYBOARD_REVE.h: add NOTE (raw ATmega, custom Makerbot board)
- pins_MIGHTYBOARD_REVG.h: add NOTE (raw ATmega, custom Makerbot board)
- pins_ORTUR_4.h: add NOTE (non-standard Arduino pin layout, no EXP tokens)
- pins_ALLIGATOR_R2.h: add NOTE (raw Due pins, UI_VOLTAGE_LEVEL board-specific)
- pins_KRATOS32.h: add NOTE (raw Due pins, BTN_BACK board-specific extra)

thinkyhead committed:
Pins: Migrate ZONESTAR_LCD to lcd/ZONESTAR_LCD.h

Create lcd/ZONESTAR_LCD.h for the standard EXP1 layout used by BTT E3-family
and FLY RRF E3 boards. Wire via pins_lcd.h.

- 8 boards cleaned: BTT_SKR_E3_TURBO, BTT_SKR_E3_DIP, BTT_SKR_MINI_E3_common,
  BTT_E3_RRF, BTT_SKR_MINI_E3_V3_0_1, FLY_RRF_E3_V1, BTT_MANTA_E3_EZ_V1_0,
  BTT_SKR_MINI_E3_V3_0: LCD_PINS_* removed; CONTROLLER_WARNING + ADC_KEYPAD_PIN retained.
- 7 boards annotated with NOTE comments (kept inline):
  RAMPS/RAMPS_144 (AUX2 layout), ZRIB_V20 (raw pins + LCD_PINS_DEFINED sentinel),
  ZRIB_V53 (non-standard EXP1 slots), DUPLICATOR_I3_PLUS (raw pins),
  SANGUINOLOLU_11 (AUX1 layout), ADSK (custom 3.3V/5V ADC scaling macros)

thinkyhead committed:
Pins: Migrate ELB_FULL_GRAPHIC_CONTROLLER to lcd/

Create lcd/ELB_FULL_GRAPHIC_CONTROLLER.h with standard EXP1/EXP2 wiring.
Uses #ifndef guards for BEEPER_PIN, SD_DETECT_PIN, KILL_PIN to allow
board-specific overrides. Wire via pins_lcd.h.

- 3 boards cleaned (inline block removed): RAMPS_NATIVE, RAMPS_144, RAMPS
- 3 boards annotated NOTE keep-inline:
  TT_OSCAR, RP2040 (raw Arduino pin numbers, not EXP-token based)
  CHITU3D (raw STM32 port pins, unique custom wiring)
- 6 files already clean (all-commented TO TEST stubs or #error only)

thinkyhead committed:
Marlin Pins: Migrate IS_TFTGLCD_PANEL to pins/lcd/TFTGLCD.h

- Create Marlin/src/pins/lcd/TFTGLCD.h with #ifndef-guarded default TFTGLCD_CS = EXP2_03_PIN
- Add #elif IS_TFTGLCD_PANEL / #include "lcd/TFTGLCD.h" entry to pins_lcd.h
- Clean 27 boards with standard EXP2_03 CS: removed inline block, replaced with "// Migrated to pins/lcd"
- Boards retaining SD_DETECT_PIN/KILL_PIN/BEEPER_PIN quirks keep slim inline block alongside the migrated comment
- 13 boards keep inline TFTGLCD_CS with NOTE comments explaining non-standard wiring:
  EXP1_03: BTT_SKR_MINI_E3_common, BTT_SKR_MINI_E3_V3_0, BTT_SKR_MINI_E3_V3_0_1, BTT_E3_RRF, FLY_RRF_E3_V1
  EXP2_05: RAMPS, RAMPS_144
  EXP2_08: XTLW_CLIMBER_8TH
  Raw pins: RAMPS_RE_ARM (P3_26), SMOOTHIEBOARD (P3_26), MKS_ROBIN_LITE3 (PB11), MKS_ROBIN_PRO (PG5), MORPHEUS (PA4)

thinkyhead committed:
MKS_12864OLED by sonnet

thinkyhead committed:
Pins: Add FYSETC_MINI_12864 inline-retention NOTEs; fix MKS_12864OLED comment

thinkyhead committed:
♻️ Migrate FYSETC_MINI_12864 standard-EXP boards

thinkyhead committed:
Pins: Document FYSETC_MINI_12864 inline retention in non-standard boards

thinkyhead committed:
♻️ Migrate MKS_MINI_12864 standard-EXP boards; annotate non-standard keepers

thinkyhead committed:
format
